crates/wasm-encoder/src/core/dump.rs

```diff
@@ -0,0 +1,228 @@
+use std::borrow::Cow;
+
+use crate::{CustomSection, Encode, Section};
+
+/// The "core" custom section for coredumps, as described in the
+/// [tool-conventions
+/// repository](https://github.com/WebAssembly/tool-conventions/blob/main/Coredump.md)
+///
+/// # Example
+///
+/// ```
+/// use wasm_encoder::{CoreDumpSection, Module};
+/// let core = CoreDumpSection::new("MyModule.wasm");
+/// let mut module = Module::new();
+/// module.section(&core);
+/// ```
+#[derive(Clone, Debug, Default)]
+pub struct CoreDumpSection {
+    name: String,
+}
+
+impl CoreDumpSection {
+    /// Create a new core dump section encoder
+    pub fn new(name: impl Into<String>) -> Self {
+        let name = name.into();
+        CoreDumpSection { name }
+    }
+}
+
+impl Encode for CoreDumpSection {
+    fn encode(&self, sink: &mut Vec<u8>) {
+        let mut data = vec![0];
+        self.name.encode(&mut data);
+        CustomSection {
+            name: "core".into(),
+            data: Cow::Owned(data),
+        }
+        .encode(sink);
+    }
+}
+
+impl Section for CoreDumpSection {
+    fn id(&self) -> u8 {
+        crate::core::SectionId::Custom as u8
+    }
+}
+
+/// A "corestack" custom section as described in the [tool-conventions
+/// repository](https://github.com/WebAssembly/tool-conventions/blob/main/Coredump.md)
+///
+/// Example
+///
+/// ```
+/// use wasm_encoder::{CoreDumpStackSection, Module, CoreDumpValue};
+/// let mut thread = CoreDumpStackSection::new("main");
+/// thread.frame(42, 1000, vec![CoreDumpValue::I32(1)], vec![CoreDumpValue::I32(2)]);
```

Comment:
Nitpick: to make this more immediately-grokkable to API docs readers, lets name these arguments:

```suggestion
///
/// let func_index = 42;
/// let code_offset = 0x1234;
/// let locals = vec![CoreDumpValue::I32(1)];
/// let stack = vec![CoreDumpValue::I32(2)];
/// thread.frame(func_index, code_offset, locals, stack);
///
```